helper/max_length_slice.go

```diff
@@ -0,0 +1,64 @@
+package helper
+
+import "sync"
+
+// MaxLengthStringSliceAdaptor wrap the given slice with several methods for MaxLenSlice.
+// Note that this adaptor does not own that slice at all, so keep the content of slice unchanged!
+type MaxLengthStringSliceAdaptor struct {
+	s      []string
+	maxlen int
+	lock   sync.Mutex
+}
+
+// NewMaxLengthStringSliceAdaptor creates an adaptor for given string slice
+func NewMaxLengthStringSliceAdaptor(s []string, maxlen int) *MaxLengthStringSliceAdaptor {
+	result := &MaxLengthStringSliceAdaptor{
+		s:      s,
+		maxlen: maxlen,
+	}
+	result.removeExceededItems()
+	return result
+}
+
+// MaxLen returns the maximum length of given maxlenslice
+func (m MaxLengthStringSliceAdaptor) MaxLen() int {
+	return m.maxlen
+}
+
+// Use it when you acquire the lock
+func (m *MaxLengthStringSliceAdaptor) removeExceededItems() {
+	if len(m.s) > m.MaxLen() {
+		m.s = m.s[len(m.s)-m.MaxLen():]
```

Comment:
Don't do this, make the slice cycle instead of create new slice. The golang's slice reuses an inner array of the original, that is it works much like the substring method of Java before 1.8. The memory may not be released correctly as your expect!